sled-agent/src/artifact_store.rs

```diff
@@ -0,0 +1,591 @@
+                        &self.log,
+                        "Failed to open file";
+                        "error" => &err,
+                        "path" => path.as_str(),
+                    );
+                    last_error = Some(Error::File { verb: "open", path, err });
+                }
+            }
+        }
+        if let Some(last_error) = last_error {
+            Err(last_error)
+        } else {
+            Err(Error::NotFound { sha256 })
+        }
+    }
+
+    /// PUT operation (served by Sled Agent API)
+    pub(crate) async fn put(
+        &self,
+        sha256: ArtifactHash,
+        stream: impl Stream<Item = Result<impl AsRef<[u8]>, Error>>,
+    ) -> Result<(), Error> {
+        let mountpoint = self
+            .dataset_mountpoints()
+            .await?
+            .next()
+            .ok_or(Error::NoUpdateDataset)?;
+        let final_path = mountpoint.join(sha256.to_string());
+
+        let temp_dir = mountpoint.join(TEMP_SUBDIR);
+        if let Err(err) = tokio::fs::create_dir(&temp_dir).await {
+            if err.kind() != ErrorKind::AlreadyExists {
+                return Err(Error::File {
+                    verb: "create directory",
+                    path: temp_dir,
+                    err,
+                });
+            }
+        }
+        let (file, temp_path) = tokio::task::spawn_blocking(move || {
+            NamedUtf8TempFile::new_in(&temp_dir).map_err(|err| Error::File {
+                verb: "create temporary file in",
+                path: temp_dir,
+                err,
+            })
+        })
+        .await??
+        .into_parts();
+        let file =
+            NamedUtf8TempFile::from_parts(File::from_std(file), temp_path);
+
+        let (mut file, hasher) = stream
+            .try_fold(
+                (file, Sha256::new()),
+                |(mut file, mut hasher), chunk| async move {
+                    hasher.update(&chunk);
+                    match file.as_file_mut().write_all(chunk.as_ref()).await {
+                        Ok(()) => Ok((file, hasher)),
+                        Err(err) => Err(Error::File {
+                            verb: "write to",
+                            path: file.path().to_owned(),
+                            err,
+                        }),
+                    }
+                },
+            )
+            .await?;
+        file.as_file_mut().sync_data().await.map_err(|err| Error::File {
+            verb: "sync",
+            path: final_path.clone(),
+            err,
+        })?;
+
+        let digest = hasher.finalize();
+        if digest.as_slice() != sha256.as_ref() {
+            return Err(Error::HashMismatch {
+                expected: sha256,
+                actual: ArtifactHash(digest.into()),
+            });
+        }
+
+        let moved_final_path = final_path.clone();
+        tokio::task::spawn_blocking(move || {
+            file.persist(&moved_final_path).map_err(|err| Error::File {
```

Comment:
We should probably also fsync the directory that we're putting it into.

Reply:
Hmm. It's kind of surprising that the `tempfile` crate doesn't do this (https://github.com/Stebalien/tempfile/issues/110), and that the Rust standard library doesn't seem to have a way to fsync a directory either. (I wonder also if there's other calls to `tempfile`'s `persist` function in the rest of Omicron that could use an fsync.)

Reply:
(This might not matter, because as part of also wanting everything to be idempotent, I may end up writing something else to manage these in-progress files...)

Reply:
> the Rust standard library doesn't seem to have a way to fsync a directory either

Okay apparently the way you do this, per the [atomicwrites](https://github.com/untitaker/rust-atomicwrites/blob/c1b223ad040939501df1a98aeaf19f86f49a7302/src/lib.rs#L179) crate, is by opening the directory with `File::open` (?!) and then calling its `.sync_all()` method. This is now done.